File: MattEland.Emergence.LevelGen/LevelData.fs
module MattEland.Emergence.Domain.LevelData

open Newtonsoft.Json
open MattEland.Emergence.Model

type LevelInstruction (prefabId: string, x: int, y: int, encounterSet: string) =
  member this.PrefabId = prefabId;
  member this.X = x;
  member this.Y = y;
  member this.EncounterSet = encounterSet;
    
/// Represents a room that can be applied to a level
type LevelData (name: string, start: Position, instructions: LevelInstruction seq) =
  member this.Name = name
  member this.Instructions = instructions
  member this.PlayerStart: Position = start

/// Creates a level from the provided JSON
let loadDataFromJson json : LevelData = JsonConvert.DeserializeObject<LevelData> json

File: MattEland.Emergence.LevelData/LevelData.cs
using System.Collections;
using System.Collections.Generic;
using MattEland.Emergence.Model;
using Newtonsoft.Json;

namespace MattEland.Emergence.LevelData
{
    public class LevelData
    {
        public string Name { get; }
        public IEnumerable<LevelInstruction> Instructions { get; }
        public Position PlayerStart { get; }

        public LevelData(string name, Position start, IEnumerable<LevelInstruction> instructions)
        {
            Name = name;
            Instructions = instructions;
            PlayerStart = start;
        }

        public static LevelData LoadFromJson(string json) => JsonConvert.DeserializeObject<LevelData>(json);
    }

}